### ISSUE TITLE
Extract tail method

### ISSUE THREAD

**Issue by yossigil** [SpartanRefactoring/Spartanizer#255]
When you find that the suffix of a method has a tail, extract the tail passing to it only live arguments.  Use the same method name.

**Comment by yossigil** [comment]
Another example, from 
```java
   private static String runScript(final String pathname) throws IOException {
    final ProcessBuilder builder = new ProcessBuilder("/bin/bash");
    builder.command(script, pathname);
    builder.redirectErrorStream(true);
    final Process process = builder.start();
    final InputStream stdout = process.getInputStream();
    try (final BufferedReader r = new BufferedReader(new InputStreamReader(stdout))) {
      String ¢;
      for (final StringBuffer $ = new StringBuffer();; $.append(¢))
        if ((¢ = r.readLine()) == null)
          return $ + "";
    }
  }
```
to
```java
  private static String runScript(final String pathname) throws IOException {
    final ProcessBuilder builder = new ProcessBuilder("/bin/bash");
    builder.command(script, pathname);
    builder.redirectErrorStream(true);
    final Process process = builder.start();
    return runScript(process);
  }
```

followed by the automatically generated auxiliary method
```java
  public static String runScript(final Process p) throws IOException {
    try (final InputStream s = p.getInputStream();
        final BufferedReader r = new BufferedReader(new InputStreamReader(s))) {
      String ¢;
      for (final StringBuffer $ = new StringBuffer();; $.append(¢))
        if ((¢ = r.readLine()) == null)
          return $ + "";
    }
  }
```

Note that this auxiliary method must have  the same name, the parameter names could be just generic names, or the names of the variables.

**Comment by OriRoth** [comment]
Wow looks great. Can I have it? seams really usable. The existing refactoring service of eclipse does that but not automatically.
I believe there should be a heuristic involved - extracting into method is good if done early in method body (well at least not at the very end).

**Comment by yossigil** [comment]
You got it

**Comment by OriRoth** [comment]
See new tipper `ExtractMethodSuffix`. Note there are still no:
* proper documentation (algorithm description missing)
* tests

I'll finish it soon. Also I need a category for it (suggestions?).
Also it causes 2 more failing tests (depending `MINIMAL_STATEMENTS_COUNT`, which is currently 6).
The basic heuristic is to extract from method of size (in statements) >= 6 and point of extraction is at the first 2/3 statements (editable, of course).
The example above works fine (in two steps - extraction then inlining, see examples below).
@yossigil closable?
Some more examples:
```
/**
 * Cool function
 * @param i
 */
void f(int i) {
  System.out.println(i);
  System.out.println(i);
  System.out.println(i);
  String j = i + "";
  System.out.println(j);
  System.out.println(j);
  System.out.println(j);
}
```
==>
```
/** 
 * Cool function
 * @param i
 */
void f(int i){
  System.out.println(i);
  System.out.println(i);
  System.out.println(i);
  String j=i + "";
  f(j);
}

/** 
 * Cool function
 * @param j
 */
void f(String j){
  System.out.println(j);
  System.out.println(j);
  System.out.println(j);
}
```
==>
```
  /** 
   * Cool function
   * @param i
   */
  void f(int i){
    System.out.println(i);
    System.out.println(i);
    System.out.println(i);
    f((i + "")); // no need for parenthesis, but not my fault
  }
  
  /** 
   * Cool function
   * @param j
   */
  void f(String j){
    System.out.println(j);
    System.out.println(j);
    System.out.println(j);
  }
```
###################################
```
void h(int ¢) {
  System.out.println(¢);
  System.out.println(¢);
  System.out.println(¢);
  System.out.println("j");
  System.out.println("j");
  System.out.println("j");
}
```
==>
```
void h(int ¢){
  System.out.println(¢);
  System.out.println(¢);
  System.out.println(¢);
  h();
}

void h(){
  System.out.println("j");
  System.out.println("j");
  System.out.println("j");
}
```
#############################
```
void g(int i) {
  System.out.println(i);
  System.out.println(i);
  System.out.println(i);
  int j = i;
  System.out.println(j);
  System.out.println(j);
  System.out.println(j);
}
```
==>
```
void g(int i){
  System.out.println(i);
  System.out.println(i);
  System.out.println(i);
  int j=i;
  g(j);
}

void g2(int j){
  System.out.println(j);
  System.out.println(j);
  System.out.println(j);
}
```
==>
```
void g(int i){
  System.out.println(i);
  System.out.println(i);
  System.out.println(i);
  g(i);
}

void g2(int j){
  System.out.println(j);
  System.out.println(j);
  System.out.println(j);
}
```
#####################################
```
int x(int a, int b) {
  System.out.println(a);
  System.out.println(a);
  System.out.println(a);
  String c = a + "";
  System.out.println(c);
  System.out.println(c);
  System.out.println(c);
  return 0;
}
```
==>
```
int x(int a,int b){
  System.out.println(a);
  System.out.println(a);
  System.out.println(a);
  String c=a + "";
  return x(c);
}

int x(String c){
  System.out.println(c);
  System.out.println(c);
  System.out.println(c);
  return 0;
}
```
==>
```
int x(int a,int b){
  System.out.println(a);
  System.out.println(a);
  System.out.println(a);
  return x((a + ""));
  }
  
int x(String c){
  System.out.println(c);
  System.out.println(c);
  System.out.println(c);
  return 0;
}
```

**Comment by yossigil** [comment]
Do you know that Laconize all stopped working, so did apply to entire project?

**Comment by OriRoth** [comment]
I'll check it out

**Comment by OriRoth** [comment]
OK try to pull, `LaconizeAll` seams to work.
I'll try to check what crashes ExtractMethodSuffix... there is some work to do, but you can check it's operation in a debugging environment.

**Comment by M4tt30rru** [comment]
I am sorry to intrude (I will delete this post afterwords). @yossigil I need a quick feedback about SAC.

**Comment by OriRoth** [comment]
Some fixes were made, now both projects JUnit and Spartanizer won't crash with `ExtractMethodSuffix`. Should I reintroduce the tipper to `ToolBox`?

**Comment by yossigil** [comment]
Sure. Good job.

**Comment by OriRoth** [comment]
Added logic to support multiple variables in extracted method. Example:
```
R f(A a, B b, C c) {
  X x = processA(a);
  YZ y = processB(b), z = processC(c);
  useX(x);
  useYZ(y);
  useYZ(z);
  return R.of(x, y, z);
}
```
==>
```
  R f(A a, B b, C c) {
    return f(processA(a), processB(b), processC(c));
  }

  R f(X x, YZ y, YZ z) {
    useX(x);
    useYZ(y);
    useYZ(z);
    return R.of(x, y, z);
  }
```
I'll upload the tipper once tests and documentation will be written.

**Comment by yossigil** [comment]
Here is something for you to try:
```java
package gnu.getopt;

import java.text.*;
import java.util.*;

public class Getopt extends Object {
  protected static final int REQUIRE_ORDER = 1;
  protected static final int PERMUTE = 2;
  protected static final int RETURN_IN_ORDER = 3;
  protected String optarg;
  protected int optind;
  protected boolean opterr = true;
  protected int optopt = '?';
  protected String nextchar;
  protected String optstring;
  protected LongOpt[] long_options;
  protected boolean long_only;
  protected int longind;
  protected boolean posixly_correct;
  protected boolean longopt_handled;
  protected int first_nonopt = 1;
  protected int last_nonopt = 1;
  private boolean endparse;
  protected String[] argv;
  protected int ordering;
  protected String progname;
  private ResourceBundle _messages = ResourceBundle.getBundle("gnu/getopt/MessagesBundle", Locale.getDefault());

  public Getopt(final String progname, final String[] argv, final String optstring) {
    this(progname, argv, optstring, null, false);
  }

  public Getopt(final String progname, final String[] argv, final String optstring, final LongOpt[] long_options) {
    this(progname, argv, optstring, long_options, false);
  }

  public Getopt(final String progname, final String[] argv, String optstring, final LongOpt[] long_options, final boolean long_only) {
    if (optstring.length() == 0)
      optstring = " ";
    this.progname = progname;
    this.argv = argv;
    this.optstring = optstring;
    this.long_options = long_options;
    this.long_only = long_only;
    if (System.getProperty("gnu.posixly_correct", null) == null)
      posixly_correct = false;
    else {
      posixly_correct = true;
      _messages = ResourceBundle.getBundle("gnu/getopt/MessagesBundle", Locale.US);
    }
    if (optstring.charAt(0) == '-') {
      ordering = RETURN_IN_ORDER;
      if (optstring.length() > 1)
        this.optstring = optstring.substring(1);
    } else if (optstring.charAt(0) != '+')
      ordering = !posixly_correct ? PERMUTE : REQUIRE_ORDER;
    else {
      ordering = REQUIRE_ORDER;
      if (optstring.length() > 1)
        this.optstring = optstring.substring(1);
    }
  }

  public void setOptstring(String optstring) {
    if (optstring.length() == 0)
      optstring = " ";
    this.optstring = optstring;
  }

  public int getOptind() {
    return optind;
  }

  public void setOptind(final int optind) {
    this.optind = optind;
  }

  public void setArgv(final String[] argv) {
    this.argv = argv;
  }

  public String getOptarg() {
    return optarg;
  }

  public void setOpterr(final boolean opterr) {
    this.opterr = opterr;
  }

  public int getOptopt() {
    return optopt;
  }

  public int getLongind() {
    return longind;
  }

  protected void exchange(final String[] argv) {
    int bottom = first_nonopt;
    final int middle = last_nonopt;
    int top = optind;
    while (top > middle && middle > bottom)
      if (top - middle <= middle - bottom) {
        final int len = top - middle;
        for (int ¢ = 0; ¢ < len; ++¢)
          swap(argv, ¢ + bottom, ¢ + middle);
        bottom = len;
      } else {
        final int len = middle - bottom;
        for (int ¢ = 0; ¢ < len; ++¢)
          swap(argv, ¢ + bottom, ¢ + top + bottom - middle);
        top -= len;
      }
    first_nonopt = optind - last_nonopt;
    last_nonopt = optind;
  }

  public static void swap(final String[] argv, final int j, final int i) {
    final String tem = argv[j];
    argv[j] = argv[i];
    argv[i] = tem;
  }

  protected int checkLongOption() {
    LongOpt pfound = null;
    int nameend;
    boolean ambig;
    boolean exact;
    longopt_handled = true;
    exact = ambig = false;
    longind = -1;
    nameend = nextchar.indexOf("=");
    if (nameend == -1)
      nameend = nextchar.length();
    for (int ¢ = 0; ¢ < long_options.length; ++¢)
      if (long_options[¢].getName().startsWith(nextchar.substring(0, nameend))) {
        if (long_options[¢].getName().equals(nextchar.substring(0, nameend))) {
          pfound = long_options[¢];
          longind = ¢;
          exact = true;
          break;
        }
        if (pfound != null)
          ambig = true;
        else {
          pfound = long_options[¢];
          longind = ¢;
        }
      }
    if (ambig && !exact) {
      if (opterr) {
        final Object[] msgArgs = { progname, argv[optind] };
        System.err.println(MessageFormat.format(_messages.getString("getopt.ambigious"), msgArgs));
      }
      nextchar = "";
      optopt = 0;
      ++optind;
      return '?';
    }
    if (pfound != null) {
      ++optind;
      if (nameend == nextchar.length()) {
        if (pfound.has_arg == LongOpt.REQUIRED_ARGUMENT) {
          if (optind >= argv.length) {
            if (opterr) {
              final Object[] msgArgs = { progname, argv[optind - 1] };
              System.err.println(MessageFormat.format(_messages.getString("getopt.requires"), msgArgs));
            }
            nextchar = "";
            optopt = pfound.val;
            return optstring.charAt(0) == ':' ? ':' : '?';
          }
          optarg = argv[optind];
          ++optind;
        }
      } else {
        if (pfound.has_arg == LongOpt.NO_ARGUMENT) {
          if (opterr)
            if (argv[optind - 1].startsWith("--")) {
              final Object[] msgArgs = { progname, pfound.name };
              System.err.println(MessageFormat.format(_messages.getString("getopt.arguments1"), msgArgs));
            } else {
              final Object[] msgArgs = { progname, Character.valueOf(argv[optind - 1].charAt(0)) + "", pfound.name };
              System.err.println(MessageFormat.format(_messages.getString("getopt.arguments2"), msgArgs));
            }
          nextchar = "";
          optopt = pfound.val + 1;
          return '?';
        }
        optarg = nextchar.substring(nameend).length() <= 1 ? "" : nextchar.substring(nameend + 1);
      }
      nextchar = "";
      if (pfound.flag == null)
        return pfound.val;
      pfound.flag.setLength(0);
      pfound.flag.append(pfound.val);
      return 0;
    }
    longopt_handled = false;
    return 0;
  }

  public int getopt() {
    optarg = null;
    if (endparse)
      return -1;
    if (nextchar == null || "".equals(nextchar)) {
      if (last_nonopt > optind)
        last_nonopt = optind;
      if (first_nonopt > optind)
        first_nonopt = optind;
      if (ordering == PERMUTE) {
        if (first_nonopt != last_nonopt && last_nonopt != optind)
          exchange(argv);
        else if (last_nonopt != optind)
          first_nonopt = optind;
        while (optind < argv.length && ("".equals(argv[optind]) || argv[optind].charAt(0) != '-' || "-".equals(argv[optind])))
          ++optind;
        last_nonopt = optind;
      }
      if (optind != argv.length && "--".equals(argv[optind])) {
        ++optind;
        if (first_nonopt != last_nonopt && last_nonopt != optind)
          exchange(argv);
        else if (first_nonopt == last_nonopt)
          first_nonopt = optind;
        optind = last_nonopt = argv.length;
      }
      if (optind == argv.length) {
        if (first_nonopt != last_nonopt)
          optind = first_nonopt;
        return -1;
      }
      if ("".equals(argv[optind]) || argv[optind].charAt(0) != '-' || "-".equals(argv[optind])) {
        if (ordering == REQUIRE_ORDER)
          return -1;
        optarg = argv[optind++];
        return 1;
      }
      nextchar = argv[optind].substring(argv[optind].startsWith("--") ? 2 : 1);
    }
    if (long_options != null
        && (argv[optind].startsWith("--") || long_only && (argv[optind].length() > 2 || optstring.indexOf(argv[optind].charAt(1)) == -1))) {
      final int $ = checkLongOption();
      if (longopt_handled)
        return $;
      if (!long_only || argv[optind].startsWith("--") || optstring.indexOf(nextchar.charAt(0)) == -1) {
        if (opterr)
          if (argv[optind].startsWith("--")) {
            final Object[] msgArgs = { progname, nextchar };
            System.err.println(MessageFormat.format(_messages.getString("getopt.unrecognized"), msgArgs));
          } else {
            final Object[] msgArgs = { progname, Character.valueOf(argv[optind].charAt(0)) + "", nextchar };
            System.err.println(MessageFormat.format(_messages.getString("getopt.unrecognized2"), msgArgs));
          }
        nextchar = "";
        ++optind;
        optopt = 0;
        return '?';
      }
    }
    int c = nextchar.charAt(0);
    nextchar = nextchar.length() <= 1 ? "" : nextchar.substring(1);
    final String temp = optstring.indexOf(c) == -1 ? null : optstring.substring(optstring.indexOf(c));
    if ("".equals(nextchar))
      ++optind;
    if (temp == null || c == ':') {
      if (opterr) {
        final Object[] msgArgs = { progname, Character.valueOf((char) c) + "" };
        System.err.println(MessageFormat.format(_messages.getString("getopt." + (posixly_correct ? "illegal" : "invalid")), msgArgs));
      }
      optopt = c;
      return '?';
    }
    if (temp.charAt(0) == 'W' && temp.length() > 1 && temp.charAt(1) == ';') {
      if (!"".equals(nextchar))
        optarg = nextchar;
      else {
        if (optind == argv.length) {
          if (opterr) {
            final Object[] msgArgs = { progname, Character.valueOf((char) c) + "" };
            System.err.println(MessageFormat.format(_messages.getString("getopt.requires2"), msgArgs));
          }
          optopt = c;
          return optstring.charAt(0) == ':' ? ':' : '?';
        }
        optarg = nextchar = argv[optind];
      }
      c = checkLongOption();
      if (longopt_handled)
        return c;
      nextchar = null;
      ++optind;
      return 'W';
    }
    if (temp.length() > 1 && temp.charAt(1) == ':') {
      if (temp.length() > 2 && temp.charAt(2) == ':')
        if ("".equals(nextchar))
          optarg = null;
        else {
          optarg = nextchar;
          ++optind;
        }
      else if (!"".equals(nextchar)) {
        optarg = nextchar;
        ++optind;
      } else {
        if (optind == argv.length) {
          if (opterr) {
            final Object[] msgArgs = { progname, Character.valueOf((char) c) + "" };
            System.err.println(MessageFormat.format(_messages.getString("getopt.requires2"), msgArgs));
          }
          optopt = c;
          return optstring.charAt(0) == ':' ? ':' : '?';
        }
        optarg = argv[optind];
        ++optind;
        if (posixly_correct && "--".equals(optarg)) {
          if (optind == argv.length) {
            if (opterr) {
              final Object[] msgArgs = { progname, Character.valueOf((char) c) + "" };
              System.err.println(MessageFormat.format(_messages.getString("getopt.requires2"), msgArgs));
            }
            optopt = c;
            return optstring.charAt(0) == ':' ? ':' : '?';
          }
          optarg = argv[optind];
          ++optind;
          first_nonopt = optind;
          last_nonopt = argv.length;
          endparse = true;
        }
      }
      nextchar = null;
    }
    return c;
  }
}

```

**Comment by OriRoth** [comment]
@yossigil challenge *accepted*!
Examine the following suggestions by my tipper:
```
protected void exchange(final String[] argv) {
  int bottom = first_nonopt;
  final int middle = last_nonopt;
  int top = optind;
  while (top > middle && middle > bottom)
    if (top - middle <= middle - bottom) {
      final int len = top - middle;
      for (int ¢ = 0; ¢ < len; ++¢)
        swap(argv, ¢ + bottom, ¢ + middle);
      bottom = len;
    } else {
      final int len = middle - bottom;
      for (int ¢ = 0; ¢ < len; ++¢)
        swap(argv, ¢ + bottom, ¢ + top + bottom - middle);
      top -= len;
    }
  first_nonopt = optind - last_nonopt;
  last_nonopt = optind;
}
```
==>
```
protected void exchange(final String[] argv) {
  int bottom = first_nonopt;
  final int middle = last_nonopt;
  int top = optind;
  while (top > middle && middle > bottom)
    if (top - middle <= middle - bottom) {
      final int len = top - middle;
      for (int ¢ = 0; ¢ < len; ++¢)
        swap(argv, ¢ + bottom, ¢ + middle);
      bottom = len;
    } else {
      final int len = middle - bottom;
      for (int ¢ = 0; ¢ < len; ++¢)
        swap(argv, ¢ + bottom, ¢ + top + bottom - middle);
      top -= len;
    }
  exchange();
}

protected void exchange() {
  first_nonopt = optind - last_nonopt;
  last_nonopt = optind;
}
```
#######################
```
protected int checkLongOption() {
  LongOpt pfound = null;
  int nameend;
  boolean ambig;
  boolean exact;
  longopt_handled = true;
  exact = ambig = false;
  longind = -1;
  nameend = nextchar.indexOf("=");
  if (nameend == -1)
    nameend = nextchar.length();
  for (int ¢ = 0; ¢ < long_options.length; ++¢)
    if (long_options[¢].getName().startsWith(nextchar.substring(0, nameend))) {
      if (long_options[¢].getName().equals(nextchar.substring(0, nameend))) {
        pfound = long_options[¢];
        longind = ¢;
        exact = true;
        break;
      }
      if (pfound != null)
        ambig = true;
      else {
        pfound = long_options[¢];
        longind = ¢;
      }
    }
  if (ambig && !exact) {
    if (opterr) {
      final Object[] msgArgs = { progname, argv[optind] };
      System.err.println(MessageFormat.format(_messages.getString("getopt.ambigious"), msgArgs));
    }
    nextchar = "";
    optopt = 0;
    ++optind;
    return '?';
  }
  if (pfound != null) {
    ++optind;
    if (nameend == nextchar.length()) {
      if (pfound.has_arg == LongOpt.REQUIRED_ARGUMENT) {
        if (optind >= argv.length) {
          if (opterr) {
            final Object[] msgArgs = { progname, argv[optind - 1] };
            System.err.println(MessageFormat.format(_messages.getString("getopt.requires"), msgArgs));
          }
          nextchar = "";
          optopt = pfound.val;
          return optstring.charAt(0) == ':' ? ':' : '?';
        }
        optarg = argv[optind];
        ++optind;
      }
    } else {
      if (pfound.has_arg == LongOpt.NO_ARGUMENT) {
        if (opterr)
          if (argv[optind - 1].startsWith("--")) {
            final Object[] msgArgs = { progname, pfound.name };
            System.err.println(MessageFormat.format(_messages.getString("getopt.arguments1"), msgArgs));
          } else {
            final Object[] msgArgs = { progname, Character.valueOf(argv[optind - 1].charAt(0)) + "", pfound.name };
            System.err.println(MessageFormat.format(_messages.getString("getopt.arguments2"), msgArgs));
          }
        nextchar = "";
        optopt = pfound.val + 1;
        return '?';
      }
      optarg = nextchar.substring(nameend).length() <= 1 ? "" : nextchar.substring(nameend + 1);
    }
    nextchar = "";
    if (pfound.flag == null)
      return pfound.val;
    pfound.flag.setLength(0);
    pfound.flag.append(pfound.val);
    return 0;
  }
  longopt_handled = false;
  return 0;
}
```
==>
```
protected int checkLongOption() {
  LongOpt pfound = null;
  int nameend;
  boolean ambig;
  boolean exact;
  longopt_handled = true;
  exact = ambig = false;
  longind = -1;
  nameend = nextchar.indexOf("=");
  if (nameend == -1)
    nameend = nextchar.length();
  for (int ¢ = 0; ¢ < long_options.length; ++¢)
    if (long_options[¢].getName().startsWith(nextchar.substring(0, nameend))) {
      if (long_options[¢].getName().equals(nextchar.substring(0, nameend))) {
        pfound = long_options[¢];
        longind = ¢;
        exact = true;
        break;
      }
      if (pfound != null)
        ambig = true;
      else {
        pfound = long_options[¢];
        longind = ¢;
      }
    }
  return ambig && !exact ? checkLongOption2() : checkLongOption(pfound, nameend);
}

protected int checkLongOption2() {
  if (opterr) {
    final Object[] msgArgs = { progname, argv[optind] };
    System.err.println(MessageFormat.format(_messages.getString("getopt.ambigious"), msgArgs));
  }
  nextchar = "";
  optopt = 0;
  ++optind;
  return '?';
}

protected int checkLongOption(LongOpt pfound, int nameend) {
  if (pfound != null) {
    ++optind;
    if (nameend == nextchar.length()) {
      if (pfound.has_arg == LongOpt.REQUIRED_ARGUMENT) {
        if (optind >= argv.length) {
          if (opterr) {
            final Object[] msgArgs = { progname, argv[optind - 1] };
            System.err.println(MessageFormat.format(_messages.getString("getopt.requires"), msgArgs));
          }
          nextchar = "";
          optopt = pfound.val;
          return optstring.charAt(0) == ':' ? ':' : '?';
        }
        optarg = argv[optind];
        ++optind;
      }
    } else {
      if (pfound.has_arg == LongOpt.NO_ARGUMENT) {
        if (opterr)
          if (argv[optind - 1].startsWith("--")) {
            final Object[] msgArgs = { progname, pfound.name };
            System.err.println(MessageFormat.format(_messages.getString("getopt.arguments1"), msgArgs));
          } else {
            final Object[] msgArgs = { progname, Character.valueOf(argv[optind - 1].charAt(0)) + "", pfound.name };
            System.err.println(MessageFormat.format(_messages.getString("getopt.arguments2"), msgArgs));
          }
        nextchar = "";
        optopt = pfound.val + 1;
        return '?';
      }
      optarg = nextchar.substring(nameend).length() <= 1 ? "" : nextchar.substring(nameend + 1);
    }
    nextchar = "";
    if (pfound.flag == null)
      return pfound.val;
    pfound.flag.setLength(0);
    pfound.flag.append(pfound.val);
    return 0;
  }
  longopt_handled = false;
  return 0;
}
```
[second split looks promising]
This code inspired me to do some changes in the splitting heuristic (to accept more cases).
Keep in mind, regarding this tipper, that the important issue is *not* whether it split where it should to, but rather if it split where it *shouldn't* to.
Do you think the tipper should've suggested more splits? maybe in `getopt`?

**Comment by yossigil** [comment]
How about #362 ?

The above is not an example for you to hack, since we need more foundational work before we can start making heuristics.

**Comment by OriRoth** [comment]
I did not hacked it! The changes were at the logical level and were very minimal: instead of checking there are no more "active" variables at the moment, I checked at least one variable is useless from the current statement (this makes place for more splitting options, something I initially hesitated doing).
After I would upload the tipper you will able to edit the heuristic. I carefully designed the tipper to separate the heuristic from the "structural" code, so it should be done with ease.
